File: benchmark/synduce/constraints/sorted_and_indexed/count_lt.f
Inductive List = elt Int | cons {Int, List};
Inductive IDList = ielt Int | icons {Int, Int, IDList};

is_sorted =
  let aux = fix (
    \f: Int -> IDList -> Bool. \pre: Int. \xs: IDList.
    match xs with
      ielt x -> > pre x
    | icons {h, _, t} -> and (> pre h) (f h t)
    end
  ) in \xs: IDList.
  match xs with
    ielt x -> true
  | icons {h, _, t} -> aux h t
  end;

len = fix (
  \f: IDList -> Int. \xs: IDList.
  match xs with
    ielt _ -> 1
  | icons {_, _, t} -> + 1 (f t)
  end
);

len_raw = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt _ -> 1
  | cons {_, t} -> + 1 (f t)
  end
);

is_indexed = fix (
  \f: IDList -> Bool. \xs: IDList.
  match xs with
    ielt _ -> true
  | icons {_, id, t} -> and (== id (len xs)) (f t)
  end
);

add_index = fix (
  \f: List -> IDList. \xs: List.
  match xs with
    elt a -> ielt a
  | cons {h, t} -> icons {h, len_raw xs, f t}
  end
);

drop_index = fix (
  \f: IDList -> List. \xs: IDList.
  match xs with
    ielt a -> elt a
  | icons {h, _, t} -> cons {h, f t}
  end
);

@Input w: Int;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> if < x w then 1 else 0
  | cons {h, t} -> + (f t) (if < h w then 1 else -1)
  end
);

target = fix (
  \f: IDList -> Compress IDList. \xs: IDList.
  match xs with
    ielt x -> xs
  | icons {h, idx, t} -> if < h w then xs else icons {h, idx, f t}
  end
);

/*Customized generator*/
main = \xs: List. let inp = add_index xs in
  if and (is_sorted inp) (is_indexed inp) then spec (drop_index (target inp)) else 0;

File: ml-benchmark/synduce/constraints/sorted_and_indexed/count_lt.ml
type list_ = Elt of int | Cons of int * list_
type idlist = IElt of int | ICons of int * int * idlist

let is_sorted =
  let rec aux pre xs =
    match xs with
    | IElt x -> pre < x
    | ICons (h, _, t) -> (pre < h) && aux h t
  in
  fun xs ->
    match xs with
    | IElt _ -> true
    | ICons (h, _, t) -> aux h t

let rec len xs =
  match xs with
  | IElt _ -> 1
  | ICons (_, _, t) -> 1 + len t

let rec len_raw xs =
  match xs with
  | Elt _ -> 1
  | Cons (_, t) -> 1 + len_raw t

let rec is_indexed xs =
  match xs with
  | IElt _ -> true
  | ICons (_, id, t) -> (id = len xs) && is_indexed t

let rec add_index xs =
  match xs with
  | Elt a -> IElt a
  | Cons (h, t) -> ICons (h, len_raw xs, add_index t)

let rec drop_index xs =
  match xs with
  | IElt a -> Elt a
  | ICons (h, _, t) -> Cons (h, drop_index t)

let w = 0

let rec spec xs =
  match xs with
  | Elt x -> if x < w then 1 else 0
  | Cons (h, t) -> spec t + (if h < w then 1 else -1)

let rec target xs =
  match xs with
  | IElt _ -> xs
  | ICons (h, idx, t) -> if h < w then xs else ICons (h, idx, target t)

let program xs =
  let inp = add_index xs in
  if (is_sorted inp) && (is_indexed inp) then spec (drop_index (target inp)) else 0